### PR TITLE
Add `round()` functionality for modules with `Real` and `Rational` fields

### PR DESCRIPTION
**Pull request by ShriKaranHanda**
### :books: Description

Add `round()` function for `FreeModuleElement` class which uses the `round()`, `ceil()` and `floor()` functions of `RealField`, `RealDoubleField` and `RationalField`. This partially solves #33921 . 

Doing the same for matrices is trivial, but since this is my first technical PR, I would like to get some feedback on this first and once the changes for modules are accepted, I'll make the changes for matrices as well?

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/modules/free_module_element.pyx [hunk 1]
```diff
@@ -1057,6 +1057,109 @@ cdef class FreeModuleElement(Vector):   # abstract base class
...
+            TypeError: the underlying ring/field must be RealField, RealDoubleField, RationalField
+            sage: v = vector(CC, [1, 2.3])
+            sage: v.round()
+            Traceback (most recent call last):
+            ...
+            TypeError: the underlying ring/field must be RealField, RealDoubleField, RationalField
+        
+        .. NOTE::
+
+            Look at ``round()``, ``ceil()`` and ``floor()`` methods of ``RealField``, ``RealDoubleField``, and ``RationalField``.
```
Comment (mkoeppe) [inline src/sage/modules/free_module_element.pyx:1131]: use sphinx roles `:meth:`, `:class:`

File: src/sage/modules/free_module_element.pyx [hunk 2]
```diff
@@ -1057,6 +1057,109 @@ cdef class FreeModuleElement(Vector):   # abstract base class
...
+        
+        """
+
+        from sage.rings.rational_field import RationalField
+
+        type_flag = True
+        try:
+            field_name = self.base_ring().name()
+            if (len(field_name) >= 9 and field_name[0:9] != 'RealField' and field_name != 'RealDoubleField'):
```
Comment (mkoeppe) [inline src/sage/modules/free_module_element.pyx:1140]: Use `isinstance` with the classes from `sage.rings.abc` instead

Comment (mkoeppe) [reply src/sage/modules/free_module_element.pyx:1140]: Also I think more fields should be allowed here. Check the function `parent_is_real_numerical`

File: src/sage/modules/free_module_element.pyx [hunk 3]
```diff
@@ -1057,6 +1057,109 @@ cdef class FreeModuleElement(Vector):   # abstract base class
...
+            sage: v.round()
+            Traceback (most recent call last):
+            ...
+            TypeError: the underlying ring/field must be RealField, RealDoubleField, RationalField
+            sage: v = vector(CC, [1, 2.3])
+            sage: v.round()
+            Traceback (most recent call last):
+            ...
+            TypeError: the underlying ring/field must be RealField, RealDoubleField, RationalField
+
```
Comment (vneiger) [inline src/sage/modules/free_module_element.pyx:1128]: this blank line 1128 has white space

File: src/sage/modules/free_module_element.pyx [hunk 4]
```diff
@@ -1057,6 +1057,109 @@ cdef class FreeModuleElement(Vector):   # abstract base class
...
+        from sage.rings.rational_field import RationalField
+
+        type_flag = True
+        try:
+            field_name = self.base_ring().name()
+            if (len(field_name) >= 9 and field_name[0:9] != 'RealField' and field_name != 'RealDoubleField'):
+                # assert field is neither RealField or RealDoubleField
+                type_flag = False
+        except:
```
Comment (vneiger) [inline src/sage/modules/free_module_element.pyx:1143]: avoid bare except (specify exception)

**Comment by vneiger** [comment]
Besides the above fixes, here is a general comment on the approach and specifically the rounding up, down, to nearest.  Here we are rounding to integers, like the ``.round`` methods in ``RR``, ``RDF`` etc. The latter fields use different method names to distinguish between these rounding choices, and ``round`` only does the rounding based on the python ``round`` function, which rounds to nearest integer (the even one in ambiguous cases). The other choices have other methods ``ceil`` and ``floor``.

For consistency, I think that for matrices over ``RR`` and ``RDF`` we should not use ``round`` with an optional argument that will switch between the choices; it will be more intuitive for the user, and also will induce less maintaining trouble in the future, if ``round`` is simply a "matrix version" of the round method of the base field.

Now, the situation seems different for the rational field, where the ``round`` offers a rounding mode... here again, the approach that seems the cleanest to me would be to make the matrix ``round`` just a directly application of the underlying field's ``round`` method to all entries of the matrix.

**Comment by vneiger** [comment]
@karan0handa  Is there any progress, made or planned for soon, on this PR?

**Comment by karan0handa** [comment]
Won't be working on this anymore.

**Pull request closed without merging**